ir/node.h

```diff
@@ -50,6 +50,14 @@ class IndexedVector;  // IWYU pragma: keep
 
 // node interface
 class INode : public Util::IHasSourceInfo, public IHasDbPrint, public ICastable {
+ private:
+    template <class, class = void>
+    struct has_static_type_name : std::false_type {};
+
+    template <class T>
+    struct has_static_type_name<T, std::void_t<decltype(T::static_type_name())>> : std::true_type {
+    };
```

Comment:
This could be possibly useful elsewhere. I would at least make it a standalone class in `node.h`, not a nested one. Please also add a short docstring & possibly the `_v` templated variable equivalent like the std:: traits have.